### PR TITLE
Simplify the query history objects

### PR DESCRIPTION
**Pull request by aeisenberg**
The goal with this change is to simplify the query history to make it
possible to serialize and de serialize.

I am not a big fan of this approach right now since it relies on mucking
with prototypes when de-serializing.

Replace this with a description of the changes your pull request makes.

## Checklist

- [n/a] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [x] Issues have been created for any UI or other user-facing changes made by this pull request.
- [n/a] _[Maintainers only]_ If this pull request makes user-facing changes that require documentation changes, open a corresponding docs pull request in the [github/codeql](https://github.com/github/codeql/tree/main/docs/codeql/codeql-for-visual-studio-code) repo and add the `ready-for-doc-review` label there.

### CODE REVIEW

**Comment by aeisenberg** [comment]
@charisk would you be able to chat about this PR? I'm trying to work on a way to properly (de-)serialize a complex object. I have something that mostly works, but it is messy. Do you have any better ideas?

Can you look at these lines in particular? https://github.com/github/vscode-codeql/pull/1115/files#diff-63143db7fbfcfc5c1d2fc82847fb9e04bf8e51580ae0f5c240d8bf97631e900cR180-R222

File: extensions/ql-vscode/src/vscode-tests/no-workspace/query-results.test.ts [hunk 1]
```diff
@@ -237,19 +248,85 @@ describe('query-results', () => {
...
+        expect(allHistoryActual[0]).to.deep.eq(allHistory[0]);
+        const info = allHistoryActual[i];
+        if (info.completedQuery) {
+          const dbItem = info.completedQuery.query.dbItem;
+          expect(dbItem).to.deep.eq(allHistory[i].completedQuery!.query.dbItem);
+        }
+      }
+      expect(allHistoryActual.length).to.deep.eq(allHistory.length);
+    });
+  });
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/vscode-tests/no-workspace/query-results.test.ts:314]: This test will ensure that whatever we serialize can be properly deserialized, but as you can see, it's complex.

**Comment by charisk** [comment]
> @charisk would you be able to chat about this PR? I'm trying to work on a way to properly (de-)serialize a complex object. I have something that mostly works, but it is messy. Do you have any better ideas?
> 
> Can you look at these lines in particular? https://github.com/github/vscode-codeql/pull/1115/files#diff-63143db7fbfcfc5c1d2fc82847fb9e04bf8e51580ae0f5c240d8bf97631e900cR180-R222

If I'm reading the code right, I think we could use something like https://github.com/typestack/class-transformer to create the class instances for us. Happy to talk/pair when you're online.

**Comment by charisk** [comment]
> Thanks. Interesting library, though since we're working with nested bespoke objects, it looks like the only way to handle this is using decorators. The bigger complication is that we can't serialize some of the objects (eg- dbItem and config) since they rely on their connections to the global state.
> 
> Maybe I need to read more, but I'm not sure if this library is going to make things much simpler.

I'm not familiar with that part of the code (happy to have a more proper look tomorrow!) so I may be completely off here but it feels like those models are more complex than they should be. I'm not sure what you mean with "connection to the global state" though. Do we need to split data and behaviour to different places?

File: extensions/ql-vscode/src/run-queries.ts [hunk 2]
```diff
@@ -706,7 +738,7 @@ export async function createInitialQueryInfo(
     isQuickEval,
     isQuickQuery: isQuickQueryPath(queryPath),
     databaseInfo,
-    id: queryId++,
+    id: `${path.basename(queryPath)}-${nanoid()}`,
```
Comment (charisk) [inline extensions/ql-vscode/src/run-queries.ts:741]: Does queryPath always exist? Can we run a query without saving it?

File: extensions/ql-vscode/src/vscode-tests/no-workspace/run-queries.test.ts [hunk 3]
```diff
@@ -47,9 +46,13 @@ describe('run-queries', () => {
       const qs = createMockQueryServerClient();
       const mockProgress = 'progress-monitor';
       const mockCancel = 'cancel-token';
+      const mockQlProgram = {
+        mock: 'program'
+      } as unknown as QlProgram;
```
Comment (charisk) [inline extensions/ql-vscode/src/vscode-tests/no-workspace/run-queries.test.ts:51]: I'd personally make a more proper mock to help with debugging.

```ts
const mockQlProgram = {
  dbschemePath: '',
   libraryPath: [],
   queryPath: ''
} as QlProgram;
```

Feel free to ignore though!

Comment (aeisenberg) [reply extensions/ql-vscode/src/vscode-tests/no-workspace/run-queries.test.ts:51]: That's fine. I was just being lazy. :)

File: extensions/ql-vscode/src/query-results.ts [hunk 4]
```diff
@@ -174,15 +183,63 @@ export type FullCompletedQueryInfo = FullQueryInfo & {
...
+          // slurped queries do not need to be disposed
+          q.completedQuery.dispose = () => { /**/ };
+        }
+        return q;
+      });
+    } catch (e) {
+      void showAndLogErrorMessage('Error loading query history.', {
+        fullMessage: ['Error loading query history.', e.stack].join('\n'),
+      });
+      return [];
```
[the remark this answers is not in the document]
Comment (charisk) [reply extensions/ql-vscode/src/query-results.ts:222]: I think in general we should leave the caller to decide what to do with the result, but happy to leave it for now as we seem to do this a lot in the current code anyway.

File: extensions/ql-vscode/src/run-queries.ts [hunk 5]
```diff
@@ -367,7 +390,7 @@ async function checkDbschemeCompatibility(
 
     // query.program.dbschemePath is the dbscheme of the actual
     // database we're querying.
-    const dbschemeOfDb = await hash(query.program.dbschemePath);
+    const dbschemeOfDb = await hash(dbItem.contents.dbSchemeUri.fsPath);
```
[the remark this answers is not in the document]
Comment (charisk) [reply extensions/ql-vscode/src/run-queries.ts:399]: Thanks for the explanation!

**Pull request merged**